op-program/l2/oracle.go

```diff
@@ -0,0 +1,8 @@
+package l2
+
+import "github.com/ethereum/go-ethereum/common"
+
+type StateOracle interface {
+	NodeByHash(nodeHash common.Hash) ([]byte, error)
+	CodeByHash(codeHash common.Hash) ([]byte, error)
+}
```

Comment:
The interface looks good to me, but note that in the implementation we may switch to mixing the hash with type data, to avoid collisions with other types of hashes (and potentially other types of oracle data like the bootstrap inputs).
A go-doc to clarify the usage (especially the expected caller, and error conditions) would be nice.

Reply:
I _think_ it will make sense to keep this API using `common.Hash` and the implementation of it that backs off to the real oracle will be responsible for mixing the type data into the hash. That way the details of the oracle access are encapsulated here and callers don't need to care about any of the details of how the oracle is actually called (including how the type is mixed into the hash, memory mapping vs sys call vs local etc). Not hard to change if we need to though.

And yeah will add godoc.

Reply:
Done. I'm more and more leaning towards remove the error return value from this and just return `nil` to indicate that data is unavailable rather than an error but don't want to get tied up with retry logic in the fetchers until I've got something working end to end.